Content.Server/Atmos/EntitySystems/GasMinerSystem.cs

```diff
@@ -25,18 +25,35 @@ private void OnMinerUpdated(Entity<GasMinerComponent> ent, ref AtmosDeviceUpdate
         {
             var miner = ent.Comp;
 
-            if (!GetValidEnvironment(ent, out var environment))
+            if (!GetValidEnvironment(ent, out var environment) || !Transform(ent).Anchored || !miner.SpawnGas.HasValue)
             {
-                miner.Idle = true;
+                if (miner.MinerState != GasMinerState.Disabled)
+                {
+                    miner.MinerState = GasMinerState.Disabled;
+                    Dirty(ent);
+                }
                 return;
             }
 
             // SpawnAmount is declared in mol/s so to get the amount of gas we hope to mine, we have to multiply this by
             // how long we have been waiting to spawn it and further cap the number according to the miner's state.
             var toSpawn = CapSpawnAmount(ent, miner.SpawnAmount * args.dt, environment);
-            miner.Idle = toSpawn == 0;
-            if (miner.Idle || !miner.Enabled || !miner.SpawnGas.HasValue)
+
+            if (toSpawn == 0)
+            {
+                if (miner.MinerState != GasMinerState.Idle)
+                {
+                    miner.MinerState = GasMinerState.Idle;
+                    Dirty(ent);
+                }
                 return;
+            }
+
+            if (miner.MinerState != GasMinerState.Working)
+            {
+                miner.MinerState = GasMinerState.Working;
+                Dirty(ent);
+            }
```

Comment:
Something like this would be more readable I think:
```cs
private void OnMinerUpdated(Entity<GasMinerComponent> ent, ref AtmosDeviceUpdateEvent args)
{
    var oldState = ent.Comp.MinerState;
    float toSpawn;

    if (!GetValidEnvironment(ent, out var environment) || !Transform(ent).Anchored)
    {
        ent.Comp.MinerState = GasMinerState.Disabled;
    }
    // SpawnAmount is declared in mol/s so to get the amount of gas we hope to mine, we have to multiply this by
    // how long we have been waiting to spawn it and further cap the number according to the miner's state.
    else if ((toSpawn = CapSpawnAmount(ent, ent.Comp.SpawnAmount * args.dt, environment)) == 0)
    {
        ent.Comp.MinerState = GasMinerState.Idle;
    }
    else {
        ent.Comp.MinerState = GasMinerState.Working;

        // Time to mine some gas.
        var merger = new GasMixture(1) { Temperature = ent.Comp.SpawnTemperature };
        merger.SetMoles(ent.Comp.SpawnGas, toSpawn);
        _atmosphereSystem.Merge(environment, merger);
    }

    if (ent.Comp.MinerState != oldState)
    {
        Dirty(ent);
    }
}
```

Reply:
yep your right